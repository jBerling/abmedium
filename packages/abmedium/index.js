const {
  layer,
  isLayer,
  document,
  isDocument,
  disagreement,
  sym,
  sim,
  str,
  num,
  seq,
  seqItems,
  nil,
  mapping,
  valtype,
  lengthOf,
  editvalOf,
  equal,
} = require('./src/core');

const proj = require('./src/proj');
const treeOf = require('./src/tree-of');
const nodes = require('./src/nodes');
const layers = require('./src/layers');

module.exports = {
  layer,
  isLayer,
  document,
  isDocument,
  disagreement,
  sym,
  sim,
  str,
  num,
  seq,
  seqItems,
  nil,
  mapping,
  valtype,
  lengthOf,
  editvalOf,
  equal,

  proj,
  treeOf,
  nodes,
  layers,
};
