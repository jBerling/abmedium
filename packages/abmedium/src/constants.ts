export const symName = "sym";
export const numName = "num";
export const seqName = "seq";
export const strName = "str";
export const nilName = "nil";
export const refName = "ref";

export const scalarTypeNames: [
  typeof nilName,
  typeof strName,
  typeof numName,
  typeof symName,
  typeof refName
] = [nilName, strName, numName, symName, refName];
