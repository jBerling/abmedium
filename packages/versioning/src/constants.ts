export const mainDir = ".abv";
export const objectsDir = ".objects";
export const timestampsLayer = "timestamps";
export const counter = "counter";
export const head = "head";
export const viewStack = ".viewstack";
