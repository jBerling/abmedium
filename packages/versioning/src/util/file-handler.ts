import {
  readFile as pipeReadFile,
  writeFile as pipeWriteFile,
  unlink as pipeUnlink,
  readDir as pipeReadDir,
  ensureDir as pipeEnsureDir,
  glob as pipeGlob,
} from "./pipable-operators";

import {
  readFile,
  writeFile,
  unlink,
  readDir,
  ensureDir,
  glob,
} from "./creation-operators";
import { resolve } from "path";
import { FileHandler } from "./types";

export const fileHandler: FileHandler = {
  resolve,

  readFile,
  writeFile,
  unlink,
  readDir,
  ensureDir,
  glob,

  pipable: {
    readFile: pipeReadFile,
    writeFile: pipeWriteFile,
    unlink: pipeUnlink,
    readDir: pipeReadDir,
    ensureDir: pipeEnsureDir,
    glob: pipeGlob,
  },
};
