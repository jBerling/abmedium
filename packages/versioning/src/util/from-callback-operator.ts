import { Observable, Observer } from "rxjs";

type Next<Out> = (out: Out) => void;

type Callback<In, Out> = (
  sourceValue: In,
  next: Next<Out>,
  errorHandler: (err: Error | string) => void
) => void;

export default <T, R>(cb: Callback<T, R>) => (source: Observable<T>) =>
  new Observable((observer: Observer<R>) => {
    let items = 0;
    let completed = 0;
    let complete = () => {};

    const subscription = source.subscribe({
      next(x: T) {
        items += 1;
        cb(
          x,
          (out) => {
            observer.next(out);
            completed += 1;
            complete();
          },
          (error) => {
            observer.error(error);
          }
        );
      },
      error(err) {
        observer.error(err);
      },
      complete() {
        complete = () => {
          if (items === completed) observer.complete();
        };
        complete();
      },
    });

    return function teardown() {
      subscription.unsubscribe();
    };
  });
