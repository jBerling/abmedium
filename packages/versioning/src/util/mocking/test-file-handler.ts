import { Observable } from "rxjs";
import { FileHandler } from "../types";
import fromCallback from "../../util/from-callback-operator";
import { Minimatch } from "minimatch";

type Directory<T> = Record<string, string | T>;
export type TestFiles = Directory<Directory<Directory<string>>>;

const pathSegments = (path: string): string[] =>
  path.split("/").filter((p) => p);

const get = (directory: any, [head, ...rest]: string[]) => {
  if (!head) throw new Error("head was undefined");
  if (!rest || !rest.length) return directory[head];
  return get(directory[head], rest);
};

const warnAboutFilePath = (path: string) => {
  if (path.startsWith("/")) {
    console.warn("Actual file path used?", path);
  }
};

const getFile = (files: TestFiles, path: string) => {
  warnAboutFilePath(path);
  const res = get(files, pathSegments(path));
  if (!res || typeof res !== "string") {
    throw new Error("No file named " + path);
  }
  return { filename: path, data: res };
};

const getDir = (files: TestFiles, path: string) => {
  warnAboutFilePath(path);

  const res = path ? get(files, pathSegments(path)) : files;

  if (!res || typeof res !== "object") {
    throw new Error("No directory named " + path);
  }
  return res;
};

const createDir = (files: TestFiles, path: string) => {
  warnAboutFilePath(path);

  const segments = pathSegments(path);
  const parent = getDir(
    files,
    segments.slice(0, segments.length - 1).join("/")
  );

  parent[segments[segments.length - 1]] = {};
};

const createFile = (files, path: string, data: string): void => {
  warnAboutFilePath(path);

  const segments = pathSegments(path);
  const parent = getDir(
    files,
    segments.slice(0, segments.length - 1).join("/")
  );

  parent[segments[segments.length - 1]] = data;
};

const exists = (files: TestFiles, path: string): boolean => {
  warnAboutFilePath(path);

  try {
    const res = get(
      files,
      path.split("/").filter((p) => p)
    );
    return !!res;
  } catch (err) {
    return false;
  }
};

const filenames = (files: TestFiles): string[] => {
  const names: string[] = [];

  const walk = (dir: TestFiles, parentPath: string): void => {
    for (const key of Object.keys(dir)) {
      const data = dir[key];
      const path = parentPath ? parentPath + "/" + key : key;
      if (typeof data === "string") {
        names.push(path);
      } else if (typeof data === "object") {
        walk(data, path);
      }
    }
  };

  walk(files, "");

  return names;
};

export const testFileHandler = (files: TestFiles): FileHandler => ({
  resolve: (...pathSegments: string[]): string =>
    pathSegments.filter((s) => s).join("/"),

  readFile: (fileName) =>
    new Observable((subscriber) => {
      try {
        subscriber.next(getFile(files, fileName).data);
        subscriber.complete();
      } catch (err) {
        subscriber.error(err);
      }
    }),
  writeFile: (fileName, data) =>
    new Observable((subscriber) => {
      try {
        createFile(files, fileName, data);
        subscriber.next();
        subscriber.complete();
      } catch (err) {
        subscriber.error(err);
      }
    }),
  unlink: (fileName: string) =>
    new Observable((subscriber) => {
      try {
        const segments = pathSegments(fileName);
        const dirPath = segments.slice(0, segments.length - 1).join("/");
        const dir = getDir(files, dirPath);
        if (dir) dir[segments[segments.length - 1]] = undefined;
        subscriber.next();
        subscriber.complete();
      } catch (err) {
        subscriber.error(err);
      }
    }),
  readDir: (dirName) =>
    new Observable((subscriber) => {
      try {
        subscriber.next(Object.keys(getDir(files, dirName)));
        subscriber.complete();
      } catch (err) {
        subscriber.error(err);
      }
    }),
  ensureDir: (dirName) =>
    new Observable((subscriber) => {
      try {
        if (!exists(files, dirName)) createDir(files, dirName);
        subscriber.next();
        subscriber.complete();
      } catch (err) {
        subscriber.error(err);
      }
    }),
  glob: (pattern, options = {}) =>
    new Observable((subscriber) => {
      try {
        const mm = new Minimatch(pattern, options);
        subscriber.next(filenames(files).filter((fname) => mm.match(fname)));
        subscriber.complete();
      } catch (err) {
        subscriber.error(err);
      }
    }),

  pipable: {
    readFile: () =>
      fromCallback((filename, next, errorHandler) => {
        try {
          next(getFile(files, filename).data);
        } catch (err) {
          console.error(err);
          errorHandler(err);
        }
      }),

    writeFile: () =>
      fromCallback(([fileName, data], next, errorHandler) => {
        try {
          createFile(files, fileName, data);
          next();
        } catch (err) {
          console.error(err);
          errorHandler(err);
        }
      }),

    unlink: () =>
      fromCallback((fileName, next, errorHandler) => {
        try {
          const segments = pathSegments(fileName);
          const dirPath = segments.slice(0, segments.length - 1).join("/");
          const dir = getDir(files, dirPath);
          if (dir) dir[segments[segments.length - 1]] = undefined;
          next();
        } catch (err) {
          console.error(err);
          errorHandler(err);
        }
      }),

    readDir: () =>
      fromCallback((dirname, next, errorHandler) => {
        try {
          next(Object.keys(getDir(files, dirname)));
        } catch (err) {
          console.error(err);
          errorHandler(err);
        }
      }),

    ensureDir: () =>
      fromCallback((dirName, next, errorHandler) => {
        try {
          if (!exists(files, dirName)) createDir(files, dirName);
          next();
        } catch (err) {
          console.error(err);
          errorHandler(err);
        }
      }),

    glob: (options = {}) =>
      fromCallback((pattern: string, next, errorHandler) => {
        const mm = new Minimatch(pattern, options);
        try {
          next(filenames(files).filter((fname) => mm.match(fname)));
        } catch (err) {
          errorHandler(err);
        }
      }),
  },
});
