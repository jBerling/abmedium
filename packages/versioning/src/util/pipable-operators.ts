import fs from "fs";
import glob_ from "glob";
import fromCallback from "./from-callback-operator";

export const readFile = (encoding: "utf8") =>
  fromCallback<string, string>((fileName, next, errorHandler) => {
    fs.readFile(fileName, encoding, (err, data) => {
      if (err) errorHandler(err);
      else next(data);
    });
  });

export const writeFile = (encoding: "utf8") =>
  fromCallback<[string, string], void>(
    ([fileName, data], next, errorHandler) => {
      fs.writeFile(fileName, data, encoding, (err) => {
        if (err) errorHandler(err);
        else next();
      });
    }
  );

export const unlink = () =>
  fromCallback<string, void>((filename, next) => {
    fs.unlink(filename, () => next());
  });

export const readDir = () =>
  fromCallback<string, string[]>((filename, next, errorHandler) => {
    fs.readdir(filename, (err, files) => {
      if (err) errorHandler(err);
      else next(files);
    });
  });

export const ensureDir = () =>
  fromCallback<string, void>((dirname, next, errorHandler) => {
    try {
      if (fs.existsSync(dirname)) {
        next();
      } else {
        fs.mkdir(dirname, (err) => {
          if (err) errorHandler(err);
          else next();
        });
      }
    } catch (err) {
      errorHandler(err);
    }
  });

export const glob = () =>
  fromCallback<string, string[]>((filePattern, next, errorHandler) => {
    try {
      glob_(filePattern, {}, (err, matches) => {
        if (err) errorHandler(err);
        else next(matches);
      });
    } catch (err) {
      errorHandler(err);
    }
  });
