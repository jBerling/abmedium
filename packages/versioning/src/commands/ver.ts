import sha256 from "crypto-js/sha256";
import { enc } from "crypto-js";
import { Observable, of, throwError, forkJoin, combineLatest } from "rxjs";
import { Argv } from "yargs";
import { map, mergeMap, mergeAll, catchError } from "rxjs/operators";
import { seq, layer, proj, mapping } from "@abrovink/abmedium";
import defaultFileHandler from "../util/file-handler";
import { FileHandler } from "../util/types";
import {
  mainDir,
  objectsDir,
  timestampsLayer,
  counter,
  head,
  viewStack,
} from "../constants";

export const command = "ver [id]";

export const describe = "add the current value to the version history";

export const builder = (yargs): Argv<{ id: string; all: boolean }> => {
  return yargs
    .options({
      all: {
        alias: "a",
        description:
          "add all unstored values to the version histories (NOT IMPLEMENTED, JUST A PLACEHOLDER)",
      },
    })
    .positional("id", {
      describe: "the entity ID",
    });
};

// TODO move these
type LayerName = string;
type LayerWithSublayers = [LayerName, ViewStack];
type ViewStack = (LayerName | LayerWithSublayers)[];
type Layer = Object;

const last = (a: Array<any>) => a[a.length - 1];

const activeLayer = (stack: ViewStack, parentLayer: Layer): Layer => {
  const item = last(stack);
  if (!item) {
    return parentLayer;
  } else if (typeof item === "string") {
    parentLayer[item] = parentLayer[item] || layer();
    return parentLayer[item];
  } else {
    const [layerName, children] = item;
    parentLayer[layerName] = parentLayer[layerName] || layer();
    return activeLayer(children, parentLayer[layerName]);
  }
};

const ver = ({
  fileHandler: { resolve, readFile, writeFile, glob } = defaultFileHandler,
  hasher = ({ fileName, data }) =>
    sha256(fileName + ":::" + data).toString(enc.Hex) +
    "." +
    last(fileName.split(".")),
  now = () => new Date().toISOString(),
  archiveName,
  id,
}: {
  fileHandler?: FileHandler;
  hasher?: (options: { fileName: string; data: string }) => string;
  now?: () => string;
  archiveName: string;
  id: string;
}): Observable<void> => {
  const archivePath = resolve(archiveName);
  const mainDirPath = resolve(archivePath, mainDir);
  const objectsDirPath = resolve(mainDirPath, objectsDir);
  const idFileName = resolve(mainDirPath, id);

  return combineLatest(
    readFile(resolve(mainDirPath, viewStack), "utf8").pipe(
      map((data): ViewStack => JSON.parse(data))
    ),
    readFile(idFileName, "utf8").pipe(map((data) => JSON.parse(data))),
    glob(resolve(archiveName, id + ".*")).pipe(
      mergeAll(),
      mergeMap((fileName) =>
        forkJoin({ fileName: of(fileName), data: readFile(fileName, "utf8") })
      ),
      mergeMap(({ fileName, data }) => {
        const hashedName = hasher({ fileName, data });
        return forkJoin({
          hashedName: of(hashedName),
          _: writeFile(resolve(objectsDirPath, hashedName), data, "utf8"),
        });
      })
    )
  ).pipe(
    mergeMap(([stack, doc, { hashedName }]) => {
      const projected = proj(doc, stack, [timestampsLayer]);
      const { [counter]: currentCounter, [head]: previous } = projected;

      let c = currentCounter;
      let itemHandle = c++;
      let versionHandle = c++;

      doc[counter] = c;

      const m = stack.length ? mapping : (x, _) => x;

      const layr = activeLayer(stack, doc);
      layr[head] = m(itemHandle, previous);
      layr[itemHandle] = m(seq(previous, versionHandle), projected[itemHandle]);
      layr[versionHandle] = m(hashedName, projected[versionHandle]);
      layr[timestampsLayer] = layr[timestampsLayer] || layer({});
      layr[timestampsLayer][itemHandle] = m(
        now(),
        layr[timestampsLayer][itemHandle]
      );

      return writeFile(idFileName, JSON.stringify(doc, null, 4), "utf8");
    }),
    catchError((err) => {
      console.error(err);
      return throwError(err);
    })
  );
};

export const handler = async ({ id }) =>
  new Promise((resolve) =>
    ver({ archiveName: process.cwd(), id }).subscribe(resolve)
  );

export default ver;
