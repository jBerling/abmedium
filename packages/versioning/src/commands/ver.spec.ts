import { TestScheduler } from "rxjs/testing";
import { FileHandler } from "../util/types";
import { testFileHandler, TestFiles } from "../util/mocking/test-file-handler";
import ver from "./ver";
import { abDocument, num, seq, nil } from "@abrovink/abmedium";
import { mainDir, objectsDir, counter, viewStack, head } from "../constants";

const testScheduler = () =>
  new TestScheduler((actual, expected) => {
    expect(JSON.parse(JSON.stringify(actual))).toEqual(
      JSON.parse(JSON.stringify(expected))
    );
  });

const run = (fn) => () => testScheduler().run(fn);

const hasher = ({ fileName }: { fileName: string; data: string }): string =>
  `<${fileName.split(".")[0]}>.${fileName.split(".")[1]}`;

const now = () => "2020-07-20T18:36Z";

describe("abv ver", () => {
  let fileHandler: FileHandler;
  let archive: TestFiles;

  beforeEach(() => {
    archive = {
      [mainDir]: {
        [viewStack]: "[]",
        [objectsDir]: {},
        foo: JSON.stringify(
          abDocument({
            [counter]: num(0),
          }),
          null,
          4
        ),
      },
      "foo.md": "# Foo\n",
    };
    fileHandler = testFileHandler(archive);
  });

  test(
    "adds first version",
    run(({ expectObservable, flush }) => {
      expectObservable(
        ver({
          fileHandler,
          hasher,
          now,
          archiveName: "",
          id: "foo",
        })
      ).toBe("(a|)", { a: undefined });

      flush();

      expect(archive).toMatchSnapshot();
    })
  );

  test(
    "adds second version",
    run(({ expectObservable, flush }) => {
      // @ts-ignore
      archive[mainDir].foo = JSON.stringify(
        abDocument({
          [counter]: num(2),
          0: seq(nil, 1),
          1: "ver1.md",
          [head]: num(0),
        }),
        null,
        4
      );

      expectObservable(
        ver({
          fileHandler,
          hasher,
          now,
          archiveName: "",
          id: "foo",
        })
      ).toBe("(a|)", { a: undefined });

      flush();

      expect(archive).toMatchSnapshot();
    })
  );

  // test(
  //   "adds the first version ",
  //   run(({ expectObservable, flush }) => {
  //     archive[mainDir][viewStack] = '[["a", ["b", "c"]]]';

  //     expectObservable(
  //       ver({
  //         fileHandler,
  //         hasher,
  //         now,
  //         archiveName: "",
  //         id: "foo",
  //       })
  //     ).toBe("(a|)", { a: undefined });

  //     flush();

  //     expect(archive).toMatchSnapshot();
  //   })
  // );
});
