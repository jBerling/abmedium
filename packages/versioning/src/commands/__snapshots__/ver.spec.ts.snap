// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`abv ver adds first version 1`] = `
Object {
  ".abv": Object {
    ".objects": Object {
      "<foo>.md": "# Foo
",
    },
    ".viewstack": "[]",
    "foo": "{
    \\"0\\": [
        \\"seq\\",
        [
            null,
            1
        ]
    ],
    \\"1\\": \\"<foo>.md\\",
    \\"counter\\": 2,
    \\"__document__\\": true,
    \\"__layer__\\": true,
    \\"head\\": 0,
    \\"timestamps\\": {
        \\"0\\": \\"2020-07-20T18:36Z\\",
        \\"__layer__\\": true
    }
}",
  },
  "foo.md": "# Foo
",
}
`;

exports[`abv ver adds second version 1`] = `
Object {
  ".abv": Object {
    ".objects": Object {
      "<foo>.md": "# Foo
",
    },
    ".viewstack": "[]",
    "foo": "{
    \\"0\\": [
        \\"seq\\",
        [
            null,
            1
        ]
    ],
    \\"1\\": \\"ver1.md\\",
    \\"2\\": [
        \\"seq\\",
        [
            0,
            3
        ]
    ],
    \\"3\\": \\"<foo>.md\\",
    \\"counter\\": 4,
    \\"head\\": 2,
    \\"__document__\\": true,
    \\"__layer__\\": true,
    \\"timestamps\\": {
        \\"2\\": \\"2020-07-20T18:36Z\\",
        \\"__layer__\\": true
    }
}",
  },
  "foo.md": "# Foo
",
}
`;
